stdlib/universe/state_tracking.go

```diff
@@ -138,12 +139,18 @@ func createStateTrackingTransformation(id execute.DatasetID, mode execute.Accumu
 	if !ok {
 		return nil, nil, errors.Newf(codes.Internal, "invalid spec type %T", spec)
 	}
-	cache := execute.NewTableBuilderCache(a.Allocator())
+	mem := a.Allocator()
+	cache := execute.NewTableBuilderCache(mem)
 	d := execute.NewDataset(id, mode, cache)
 	t, err := NewStateTrackingTransformation(a.Context(), s, d, cache)
 	if err != nil {
 		return nil, nil, err
 	}
+
+	if feature.OptimizeStateTracking().Enabled(a.Context()) {
```

Comment:
You should move this feature flag check above the creation of the transformation. If you change `NewNarrowStateTrackingTransformation` to take in the context and spec independently, then you'll avoid creating unnecessary objects.